crates/engine/tree/src/tree/payload_processor/multiproof.rs

```diff
@@ -1010,37 +1287,90 @@ impl MultiProofTask {
     ///    * Once this message is received, on every [`MultiProofMessage::EmptyProof`] and
     ///      [`ProofResultMessage`], we check if all proofs have been processed and if there are any
     ///      pending proofs in the proof sequencer left to be revealed.
-    /// 6. This task exits after all pending proofs are processed.
+    /// 6. While running, consecutive [`MultiProofMessage::PrefetchProofs`] and
+    ///    [`MultiProofMessage::StateUpdate`] messages are batched to reduce redundant work; if a
+    ///    different message type arrives mid-batch or a batch cap is reached, it is held as
+    ///    `pending_msg` and processed on the next loop to preserve ordering.
+    /// 7. This task exits after all pending proofs are processed.
     #[instrument(
         level = "debug",
         name = "MultiProofTask::run",
         target = "engine::tree::payload_processor::multiproof",
         skip_all
     )]
     pub(crate) fn run(mut self) {
-        // TODO convert those into fields
-        let mut prefetch_proofs_requested = 0;
-        let mut state_update_proofs_requested = 0;
-        let mut proofs_processed = 0;
+        let mut ctx = MultiproofBatchCtx::new(Instant::now());
+        let mut batch_metrics = MultiproofBatchMetrics::default();
 
-        let mut updates_finished = false;
+        // Main event loop that prioritizes draining proof results, then processes batched control
+        // messages; labeled so inner loops can `break 'main` once all work is complete.
+        'main: loop {
+            trace!(target: "engine::tree::payload_processor::multiproof", "entering main channel receiving loop");
 
-        // Timestamp before the first state update or prefetch was received
-        let start = Instant::now();
+            // Always drain proof results first to maintain priority over pending messages.
+            // This prevents priority inversion where pending_msg would bypass select_biased!
+            // and starve proof result processing.
+            while let Ok(proof_result) = self.proof_result_rx.try_recv() {
```

Comment:
Was removing the crossbeam_select a perf related decision or was it required due to the logic changes?

Reply:
adressed [98f3406](https://github.com/paradigmxyz/reth/pull/20066/commits/98f3406949877db64003ef8e07c9f51b99c851e9)